tests/profiling/collector/test_stack.py

```diff
@@ -325,3 +325,28 @@ def test_collect_multiple_span_ids(tracer_and_collector):
             continue
         if child.trace_id in event.trace_ids:
             break
+
+
+def test_stress_trace_collection(tracer_and_collector):
```

Comment:
Did this test reproduce the error?

Reply:
I wish but no :(

Reply:
Actually it did reproduce another error that ought to be fixed by this test:

```
=================================== FAILURES ===================================
 
_________________________ test_stress_trace_collection _________________________
 

 
tracer_and_collector = (<ddtrace.tracer.Tracer object at 0x7efe98642b10>, StackCollector(status=<ServiceStatus.RUNNING: 'running'>, recorder=Recorder(ma... ignore_profiler=True, tracer=<ddtrace.tracer.Tracer object at 0x7efe98642b10>))
 

 
    def test_stress_trace_collection(tracer_and_collector):
 
        tracer, collector = tracer_and_collector
 
    
 
        def _trace():
 
            for _ in range(5000):
 
                with tracer.trace("hello"):
 
                    time.sleep(0.001)
 
    
 
        NB_THREADS = 50
 
    
 
        threads = []
 
        for i in range(NB_THREADS):
 
            t = threading.Thread(target=_trace)
 
            threads.append(t)
 
    
 
        for t in threads:
 
            t.start()
 
    
 
        for _ in range(10000):
 
>           collector.collect()
 

 
tests/profiling/collector/test_stack.py:352: 
 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 
ddtrace/profiling/collector/stack.pyx:462: in ddtrace.profiling.collector.stack.StackCollector.collect
 
    all_events = stack_collect(
 
ddtrace/profiling/collector/stack.pyx:328: in ddtrace.profiling.collector.stack.stack_collect
 
    spans = thread_span_links.get_active_leaf_spans_from_thread_id(tid)
 
ddtrace/profiling/collector/stack.pyx:402: in ddtrace.profiling.collector.stack._ThreadSpanLinks.get_active_leaf_spans_from_thread_id
 
    with self._lock:
 
ddtrace/profiling/collector/stack.pyx:403: in ddtrace.profiling.collector.stack._ThreadSpanLinks.get_active_leaf_spans_from_thread_id
 
    spans = set(self._thread_id_to_spans.get(thread_id, ()))
 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 

 
self = <_weakrefset.WeakSet object at 0x7efe98906cd0>
 

 
    def __iter__(self):
 
        with _IterationGuard(self):
 
>           for itemref in self.data:
 
E           RuntimeError: Set changed size during iteration
 

 
../.pyenv/versions/2.7.17/lib/python2.7/_weakrefset.py:60: RuntimeError
 ```

This does not make any sense since the `WeakSet` is locked. Digging…

Reply:
What was the issue? 🤔

Reply:
It's messy… since we're using weak references, they might disappear at any time… which include while iterating, because the GC might run at this time and delete the references.
There are multiple bug report over the years around this which hit CPython libraries. I'm not sure `WeakSet` is actually bug-free in our conditions, so I've just ditched it and rewrote a simpler set-based approach where we can control the reference chain and avoid the GC playing us.

Reply:
:+1: